### PR TITLE
Misc fixes to ADVI

### PR DESCRIPTION
**Pull request by fritzo**
- Support scalars in `log_abs_det_jacobian`
- Directly use `scale_tril` argument to `MultivariateNormal`
- Fix shaping issues with `StickBreakingTransform`
- Return a dict mapping sample site name -> value from the `advi.guide()`. This is free to compute and often useful.

## Tested
- unmarked lots of tests that were previously xfailing
- added a test with Dirichlet; it passes the shape part but xfails for another reason (possible bug in stick breaking jacobian)

### CODE REVIEW

**Comment by fritzo** [comment]
I'm hoping the ADVI interface will settle down before 0.2.0 release. If it is still in flux we may move it temporarily to pyro.contrib and move it back to core in the 0.2.1 release.

File: pyro/infer/advi.py [hunk 1]
```diff
@@ -43,9 +43,10 @@ def _setup_prototype(self, *args, **kwargs):
         # run the model so we can inspect its structure
         self.prototype_trace = poutine.block(poutine.trace(self.base_model).get_trace)(*args, **kwargs)
         self.prototype_trace = prune_subsample_sites(self.prototype_trace)
-        self.latent_dim = sum(site["value"].view(-1).size(0)
-                              for site in self.prototype_trace.nodes.values()
-                              if site["type"] == "sample" and not site["is_observed"])
+        self.latent_shapes = {name: biject_to(site["fn"].support).inv(site["value"]).shape
```
Comment (martinjankowiak) [inline pyro/infer/advi.py]: why is this necessary? can we have a comment?

Comment (fritzo) [reply pyro/infer/advi.py]: comment added.
